build-tools/packages/build-tools/src/type-test-generator/typeTestUtils.ts

```diff
@@ -0,0 +1,260 @@
+ * Licensed under the MIT License.
+ */
+
+import { existsSync } from "node:fs";
+import path from "node:path";
+import { PackageName } from "@rushstack/node-core-library";
+import { readJsonSync } from "fs-extra";
+import { Project, SourceFile } from "ts-morph";
+import { BrokenCompatTypes } from "../common/fluidRepo";
+import { PackageJson } from "../common/npmPackage";
+import { TestCaseTypeData, buildTestCase } from "../typeValidator/testGeneration";
+import { TypeData, getFullTypeName, getNodeTypeData } from "../typeValidator/typeData";
+
+/**
+ * The path to the test types directory
+ */
+const typeTestFilePath = "./src/test/types";
+
+/**
+ * Checks the package object to verify that the specified dependency exists
+ * @param packageObject - the package.json object to check for the dependency
+ * @param dependencyName - the dependency to check for in the package object
+ * @remarks Information about the previous package from the package.json is not needed,
+ * but error if it's missing since it's nice to separate errors for the dep missing here vs not installed.
+ * This ensures that a critical dependency (the previous package version) is correctly declared in the project's package.json.
+ */
+export function ensureDevDependencyExists(
+	packageObject: PackageJson,
+	dependencyName: string,
+): void {
+	const dependencyVersion = packageObject?.devDependencies?.[dependencyName];
+	if (typeof dependencyVersion !== "string") {
+		throw new Error(`Did not find devDependency in package.json`);
+	}
+}
+
+/**
+ * Fetches the path of the previous package.json or throws an error if not found.
+ * @param previousBasePath - A string representing the path to the root directory of the "previous" package dependency.
+ */
+export function tryGetPreviousPackageJsonPath(previousBasePath: string): string {
+	const previousPackageJsonPath = path.join(previousBasePath, "package.json");
+	if (!existsSync(previousPackageJsonPath)) {
+		throw new Error(`${previousPackageJsonPath} not found.`);
+	}
+	return previousPackageJsonPath;
+}
+
+/**
+ * Attempts to extract the type definition file path from the 'exports' field of a given package.json.
+ * Checks both 'import' and 'require' resolution methods to find the appropriate path.
+ * If the path is found, it is returned. Otherwise, an error is thrown.
+ * @param previousPackageJson - An object representing the previous package json
+ * @param previousBasePath - A string representing the path to the root directory of the "previous" package dependency.
+ * @returns A type definition filepath based on the appropriate export, or undefined if it cannot be found.
+ */
+export function getTypePathFromExport(
+	previousPackageJson: PackageJson,
+	previousBasePath: string,
+): string | undefined {
+	if (!previousPackageJson.exports) {
+		console.warn("The 'exports' field is missing in the package.json.");
+		return undefined;
+	}
+
+	const extractTypesPath = (exportEntry: any): string | undefined => {
```

Comment:
Should not use `any` typing. No lint complaints?

Reply:
The @fluidframework/build-tools package has terrible lint settings. All new code should be in build-cli. We're close to build-tools being only fluid-build, which is the ultimate goal state I think.